design/decouple-ceph-version.md

```diff
@@ -0,0 +1,128 @@
+will be reviewed. If the version has changed, the operator will update each of the daemons in a predictable order such as: mon, mgr, osd, rgw, mds. If the Rook upgrade requires any special steps, they will be handled as each version upgrade requires.
+
+When the cluster CRD is updated with a new Ceph version, the same idempotent orchestration is executed to evaluated desired state that needs to be applied to the cluster.
+Over time as the operator becomes smarter and more versions are supported, the custom upgrade steps will be implemented as needed.
+
+Daemons will only be restarted when necessary for the upgrade. The Rook upgrade sometimes will not require a restart of the daemons,
+depending on if the pod spec changed. The Ceph upgrade will always require a restart of the daemons. In either case, a restart will be done in an orderly manner
+with one pod at a time along with health checks as the upgrade proceeds. The upgrade will be paused if the cluster becomes unhealthy.
+
+#### Admin control of upgrades
+
+To allow more control over the upgrade, we define `upgradePolicy` settings.
+- `newCephImage`: The version of the image to start applying to the daemons specified in the `upgrade` list. After the admin is satisfied with the upgrade, he should delete this property and the `upgrade` properties from the CRD to declare that Rook should fully be running on the new version.
+- `upgrade`: A list of daemons that should be upgraded to the version `newCephImage`. The daemons include `mon`, `osd`, `mgr`, `rgw`, and `mds`. The ordering of the list will be ignored as Rook will only support ordering as it determines necessary for a version. If there are special upgrade actions in the future, they could be named and added to this list.
+- `allowUnrecognizedVersion`: If `false`, the operator would refuse to upgrade the Ceph version if it doesn't support or recognize that version. If `true`, Rook would go ahead and blindly set the image version and assume the pod specs should match `unrecognizedMajorVersion`. This would allow testing of upgrade to unreleased versions. The default is `false`.
+- `unrecognizedMajorVersion`: The version of pod spec to use when orchestrating an unrecognized version. Possible values are the names of versions that Rook understands: `luminous`, `mimic`, or `nautilus`.
+
+For example, with the settings below the operator would only upgrade the mons to mimic and other daemons would remain on luminous. When the admin is ready, he would add more daemons to the list.
+
+```yaml
+spec:
+  cephImage: ceph/ceph:v12.2.7
+  upgradePolicy:
+    newCephImage: ceph/ceph:v13.2.1
+    upgrade:
+    - mon
+```
+
+When the admin is completed with the upgrade or he is ready to allow Rook to complete the full upgrade for all daemons, he would set `cephImage: ceph/ceph:v13.2.1` and delete the `upgradePolicy`.
```

Comment:
What's the behaviour if the admin deletes the upgradePolicy, but doesn't modify cephImage, does the cluster then try and downgrade?  This feels like a user mistake that's relatively likely.

What if instead of an upgradePolicy, treated the cephImage field as the "goal" version, so that the admin would trigger an upgrade by simply modifying that field.  Then Rook could do its internal housekeeping to remember what version was in use for all the daemon types, and track which daemon types have been successfully migrated to new versions?

Reply:
I don't think we should support downgrade. The whitelist would ignore a lower version. In this case that means the cluster would stay in a partially upgraded state. 

I think what you're suggesting for the `cephImage` being the "goal" version is the intent of the design. If you set the `cephImage` to a new version, you're allowing Rook to go ahead with the full upgrade in a rolling manner. The version under the `upgradePolicy` is only intended to give more control to the admin for rolling upgrade for only the components he is ready to upgrade.

Reply:
FWIW the plan currently is to support downgrade in limited cases (mainly within stable versions, e.g., 13.2.1 to 13.2.0) but not across major versions.  The testing is not in place yet to allow this, though.

Reply:
@liewegas Can we assume that any special upgrade steps will happen across major versions? It would be very helpful to the test matrix if patch releases could be a simple update to the container version. But Rook could certainly put restrictions in place for patch release downgrades until it's supported.

Reply:
I think so.  I can't remember a case where we had special upgrade steps for a minor bugfix release. 

Eventually we'll probably have some case come up where that applies, though.  For example with the broken 12.2.6 release there were steps that you could take to help mitigate the regression that could conceivably we done by an operator...but I doubt we'll actually end up doing (or needing to do) this anytime soon.